File: atv6_algoritmo_genetico/js/src/individual.js
export default class Individual {
	constructor(x, y) {
		this.x = Number(x);
		this.y = Number(y);
		this.fitness = this.calculateFitness();
		this.mappedFitness = null;
	}

	calculateFitness() {
		return Number(
			(
				Math.sin(this.x) * Math.exp((1 - Math.cos(this.y)) ** 2) +
				Math.cos(this.y) * Math.exp((1 - Math.sin(this.x)) ** 2) +
				(this.x - this.y) ** 2
			).toFixed(2)
		);
	}

	tryToMutate() {
		const localSearch = 0.1;
		if (Math.random() < 0.05) {
			const xNeg = this.x - (this.x*localSearch);
			const xPos = this.x + (this.x*localSearch)
			this.x = Math.random() * (xPos - xNeg + 1) + xNeg
		}
		if (Math.random() < 0.05) {
			const yNeg = this.y - (this.y*localSearch);
			const yPos = this.y + (this.y*localSearch)
			this.y = Math.random() * (yPos - yNeg + 1) + yNeg
		}
	}

	toString() {
		return `x: ${this.x}, y: ${this.y}, fitness: ${this.fitness}, mappedFitness: ${this.mappedFitness}`;
	}
}

File: atv6_algoritmo_genetico/python/src/individual.py
import numpy as np

class Individual:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.fitness = None
        self.mappedFitness = None

    def __str__(self):
        return (
            "x: " + str(self.x) + "     " +
            "y: " + str(self.y) + "     " +
            "Fitness: " + str(self.fitness) + "     " +
            "Mapped Fitness: " + str(self.mappedFitness)
        )

    def calculateFitness(self, fitnessFunction):
        self.fitness = fitnessFunction(self.x, self.y)

    def calculateMapFitness(self, linearMappingFunction, i, N):
        self.mappedFitness = linearMappingFunction(i, N)

    def tryToMutate(self, local_search):
        if np.random.rand() < 0.1:
            xNeg = self.x - (self.x*local_search)
            xPos = self.x + (self.x*local_search)
            self.x = round(np.random.uniform(xNeg, xPos), 5)
        if np.random.rand() < 0.1:
            yNeg = self.y - (self.y*local_search)
            yPos = self.y + (self.y*local_search)
            self.y = round(np.random.uniform(yNeg, yPos), 5)
